examples/reordered-preconditioned-solver/reordered-preconditioned-solver.cpp

```diff
@@ -0,0 +1,111 @@
+    using real_vec = gko::matrix::Dense<RealValueType>;
+    using mtx = gko::matrix::Csr<ValueType, IndexType>;
+    using cg = gko::solver::Cg<ValueType>;
+    using bj = gko::preconditioner::Jacobi<ValueType, IndexType>;
+
+    // Print version information
+    std::cout << gko::version_info::get() << std::endl;
+
+    // Figure out where to run the code
+    if (argc == 2 && (std::string(argv[1]) == "--help")) {
+        std::cerr << "Usage: " << argv[0] << " [executor]" << std::endl;
+        std::exit(-1);
+    }
+
+    // Figure out where to run the code
+    const auto executor_string = argc >= 2 ? argv[1] : "reference";
+    std::map<std::string, std::function<std::shared_ptr<gko::Executor>()>>
+        exec_map{
+            {"omp", [] { return gko::OmpExecutor::create(); }},
+            {"cuda",
+             [] {
+                 return gko::CudaExecutor::create(0,
+                                                  gko::OmpExecutor::create());
+             }},
+            {"hip",
+             [] {
+                 return gko::HipExecutor::create(0, gko::OmpExecutor::create());
+             }},
+            {"dpcpp",
+             [] {
+                 return gko::DpcppExecutor::create(0,
+                                                   gko::OmpExecutor::create());
+             }},
+            {"reference", [] { return gko::ReferenceExecutor::create(); }}};
+
+    // executor where Ginkgo will perform the computation
+    const auto exec = exec_map.at(executor_string)();  // throws if not valid
+
+    // Read data
+    auto A = share(gko::read<mtx>(std::ifstream("data/A.mtx"), exec));
+    auto b = gko::read<vec>(std::ifstream("data/b.mtx"), exec);
+    auto x = gko::read<vec>(std::ifstream("data/x0.mtx"), exec);
+
+    auto reordering =
+        gko::experimental::reorder::Rcm<IndexType>::build().on(exec)->generate(
+            A);
+    // Permute matrix and vectors
+    auto A_reordered = share(A->permute(reordering));
+    auto b_reordered = b->permute(reordering, gko::matrix::permute_mode::rows);
+    // this reordering is not necessary, but it maps the initial guess to the
+    // unreordered case
+    auto x_reordered = x->permute(reordering, gko::matrix::permute_mode::rows);
+
+    const RealValueType reduction_factor{1e-7};
+    // Create solver factory
+    auto solver_gen =
+        cg::build()
+            .with_criteria(gko::stop::Iteration::build().with_max_iters(20u),
+                           gko::stop::ResidualNorm<ValueType>::build()
+                               .with_reduction_factor(reduction_factor))
+            // Add preconditioner, these 2 lines are the only
+            // difference from the simple solver example
+            .with_preconditioner(bj::build().with_max_block_size(8u))
+            .on(exec);
+    // Create solver
+    auto solver = solver_gen->generate(A_reordered);
+
+    // Solve system
+    solver->apply(b_reordered, x_reordered);
```

Comment:
Somehow, I thought there would be a wrapper that automatically applies the reordering to the input matrices and vectors. Was something like that planned?